core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -454,11 +465,17 @@ private[log] class Cleaner(val id: Int,
                              map: OffsetMap,
                              retainDeletes: Boolean,
                              maxLogMessageSize: Int,
+                             transactionMetadata: CleanedTransactionMetadata,
                              activePids: Map[Long, ProducerIdEntry],
                              stats: CleanerStats) {
     val logCleanerFilter = new RecordFilter {
-      def shouldRetain(recordBatch: RecordBatch, record: Record): Boolean =
-        shouldRetainMessage(source, map, retainDeletes, recordBatch, record, stats, activePids)
+      override def shouldDiscard(batch: RecordBatch): Boolean =
+        // we piggy-back on the tombstone retention logic to delay deletion of transaction markers.
+        // note that we will never delete a marker until all the records from that transaction are removed.
+        shouldDiscardBatch(batch, transactionMetadata, retainTxnMarkers = retainDeletes)
```

Comment:
Good catch. Fixing this requires some additional logic in the filter, but I guess it can't be helped.